packages/loader/container-loader/src/catchUpMonitor.ts

```diff
@@ -0,0 +1,101 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IDisposable, IEvent } from "@fluidframework/common-definitions";
+import { assert, TypedEventEmitter } from "@fluidframework/common-utils";
+import { IDeltaManager } from "@fluidframework/container-definitions";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+
+/** @see ICatchUpMonitor for usage */
+type CaughtUpListener = (hasCheckpointSequenceNumber: boolean) => void;
+
+/** @see ICatchUpMonitor for usage */
+export interface ICatchUpMonitorEvents extends IEvent {
+    (event: "caughtUp", listener: CaughtUpListener): void;
+}
+
+/** Monitor that emits an event when a Container has caught up to a given point in the op stream */
+export interface ICatchUpMonitor extends TypedEventEmitter<ICatchUpMonitorEvents>, IDisposable { }
+
+/**
+ * Monitors a Container's DeltaManager, notifying listeners when all ops have been processed
+ * that were known at the time the monitor was created.
+ */
+export class CatchUpMonitor extends TypedEventEmitter<ICatchUpMonitorEvents> implements ICatchUpMonitor {
+    private readonly hasCheckpointSequenceNumber: boolean;
+    private readonly targetSeqNumber: number;
+    private caughtUp: boolean = false;
+
+    private readonly opHandler = (message: Pick<ISequencedDocumentMessage, "sequenceNumber">) => {
+        if (!this.caughtUp && message.sequenceNumber >= this.targetSeqNumber) {
+            this.caughtUp = true;
+            this.emit("caughtUp", this.hasCheckpointSequenceNumber);
+        }
+    };
+
+    /**
+     * Create the CatchUpMonitor, setting the target sequence number to wait for based on DeltaManager's current state.
+     */
+    constructor(
+        private readonly deltaManager: IDeltaManager<any, any>,
+    ) {
+        super();
+
+        this.hasCheckpointSequenceNumber = this.deltaManager.hasCheckpointSequenceNumber;
+        this.targetSeqNumber = this.deltaManager.lastKnownSeqNumber;
+
+        assert(this.targetSeqNumber >= this.deltaManager.lastSequenceNumber,
+            "Cannot wait for seqNumber below last processed sequence number");
+
+        this.deltaManager.on("op", this.opHandler);
+
+        // Simulate the last processed op to set caughtUp in case we already are
+        this.opHandler({ sequenceNumber: this.deltaManager.lastSequenceNumber });
+
+        // If a listener is added after we are already caught up, notify that new listener immediately
+        this.on("newListener", (event: string, listener) => {
+            if (event === "caughtUp") {
+                const caughtUpListener = listener as CaughtUpListener;
+                if (this.caughtUp) {
+                    caughtUpListener(this.hasCheckpointSequenceNumber);
```

Comment:
You are passing "transitionToConnectedState" as "caughtUpListener", but transitionToConnectedState does not take any argument. So, what is the logic here of passing the "hasCheckpointSequenceNumber".

Reply:
I went back and forth on this.  In the future when this feature gate is at 100%, `waitContainerToConnect` will become obsolete, replaced by simply waiting for the `connected` event.  If someone were depending on the boolean return value of that function, maybe we would want to update `connected` to include that in its event args.

I feel it's kind of a strange thing to expose, I guess the point is some kind of "confidence" indicator of how caught up you'll be when the thing resolves/fires.  @vladsud thoughts on maintaining this info long-term (and whether it belongs in the short term)

Reply:
I'm fine not exposing this info anywhere. It used to be that FRS was an outlier here as it did not have checkpoint sequence number. But with recent work to expose sequence numbers on join & leave signals, we should ensure that both FRS is on par with ODSP, as this feature is just too good not to have it - it allows many things beyond ability to catch up, like building sync primitives (including leader / summarizer election) using signals only. Plus, it was added to solve race condition on connection that resulted in client missing ops - so it's really not that optional to have it.